File: BehavioralDesignPatterns/Java/Strategy.java
// Strategy Interface
interface Strategy {
    int execute(int a, int b);
}

// ConcreteStrategy for Addition
class AddStrategy implements Strategy {
    public int execute(int a, int b) {
        return a + b;
    }
}

// ConcreteStrategy for Subtraction
class SubtractStrategy implements Strategy {
    public int execute(int a, int b) {
        return a - b;
    }
}

// ConcreteStrategy for Multiplication
class MultiplyStrategy implements Strategy {
    public int execute(int a, int b) {
        return a * b;
    }
}

// Context Class
class Calculator {
    private Strategy strategy;

    public Calculator(Strategy strategy) {
        this.strategy = strategy;
    }

    public void setStrategy(Strategy strategy) {
        this.strategy = strategy;
    }

    public int executeStrategy(int a, int b) {
        return strategy.execute(a, b);
    }
}

// Client Code
public class Main {
    public static void main(String[] args) {
        Strategy addStrategy = new AddStrategy();
        Strategy subtractStrategy = new SubtractStrategy();
        Strategy multiplyStrategy = new MultiplyStrategy();

        Calculator calculator = new Calculator(addStrategy);
        System.out.println(calculator.executeStrategy(5, 3));  // Output: 8

        calculator.setStrategy(subtractStrategy);
        System.out.println(calculator.executeStrategy(5, 3));  // Output: 2

        calculator.setStrategy(multiplyStrategy);
        System.out.println(calculator.executeStrategy(5, 3));  // Output: 15
    }
}

File: BehavioralDesignPatterns/JavaScript/Strategy.js
// Strategy Interface
class Strategy {
    execute(a, b) {
        throw new Error("This method must be overridden.");
    }
}

// ConcreteStrategy for Addition
class AddStrategy extends Strategy {
    execute(a, b) {
        return a + b;
    }
}

// ConcreteStrategy for Subtraction
class SubtractStrategy extends Strategy {
    execute(a, b) {
        return a - b;
    }
}

// ConcreteStrategy for Multiplication
class MultiplyStrategy extends Strategy {
    execute(a, b) {
        return a * b;
    }
}

// Context Class
class Calculator {
    constructor(strategy) {
        this.strategy = strategy;
    }

    setStrategy(strategy) {
        this.strategy = strategy;
    }

    executeStrategy(a, b) {
        return this.strategy.execute(a, b);
    }
}

// Client Code
const addStrategy = new AddStrategy();
const subtractStrategy = new SubtractStrategy();
const multiplyStrategy = new MultiplyStrategy();

const calculator = new Calculator(addStrategy);
console.log(calculator.executeStrategy(5, 3));  // Output: 8

calculator.setStrategy(subtractStrategy);
console.log(calculator.executeStrategy(5, 3));  // Output: 2

calculator.setStrategy(multiplyStrategy);
console.log(calculator.executeStrategy(5, 3));  // Output: 15
